elasticsearch/class-health-command.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace Automattic\VIP\Elasticsearch;
+
+use \WP_CLI;
+use \WP_CLI\Utils;
+use \WP_Query as WP_Query;
+use \ElasticPress\Indexables as Indexables;
+use \ElasticPress\Elasticsearch as Elasticsearch;
+
+/**
+ * Helper commands to manage VIP Go Elasticsearch indexes
+ *
+ * @package Automattic\VIP\Elasticsearch
+ */
+class Health_Command extends \WPCOM_VIP_CLI_Command {
+
+	/**
+	 * @var WP_CLI\Fetchers\Plugin Plugin fetcher
+	 */
+	protected $plugin_fetcher;
+
+	public function __construct() {
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		parent::__construct();
+
+		$this->plugin_fetcher = new \WP_CLI\Fetchers\Plugin();
+	}
+
+	/**
+	 * Validate DB and ES index post counts
+	 *
+	 * ## OPTIONS
+	 *
+	 *
+	 * ## EXAMPLES
+	 *     wp vip-es health validate-counts
+	 *
+	 * @subcommand validate-counts
+	 */
+	public function validate_counts( $args, $assoc_args ) {
+		// Make sure ElasticPress is installed and active
+		$this->is_elasticpress_active();
+
+		// Get indexable objects
+		$indexable = Indexables::factory()->get( 'post' );
+
+		$post_types = $indexable->get_indexable_post_types();
+
+		WP_CLI::line( sprintf( "Checking %d post types (%s)\n", count( $post_types ), implode( ',', $post_types ) ) );
+
+		$error = false;
+		$es_conn_err = false;
+
+		foreach( $post_types as $post_type ) {
+			$query_args = [ 
+				'post_type' => $post_type,
+			];
+
+			// Get total count in DB
+			$result = $indexable->query_db( $query_args );
+
+			$db_total = (int) $result[ 'total_objects' ];
+
+			// Get total count in ES index
+			$query = new WP_Query( $query_args );
+
+			$formatted_args = $indexable->format_args( $query->query_vars, $query );
+
+			$es_result = $indexable->query_es( $formatted_args, $query->query_vars );
+			
+			if ( ! $es_result ) {
+				$es_total = 'N/A';
+				$error = true;
+				$es_conn_err = true;
```

Comment:
Should this condition also break the `foreach()`? Or, maybe we should just put the logic from line 81 into the `else` statement, so that it doesn't try to run after an error.

Reply:
I didn't break the `foreach()` loop as I'm thinking it could only be temporary and the number of post types could vary depending on the site.

Good point about putting that logic in the `else` statement though.

Reply:
Changed in 21a5ac773c63707f872a89c74b79b56698fa3683